### PR TITLE
Additional API to change model TXD

### PR DESCRIPTION
**Pull request by TheNormalnij**
```
bool engineSetModelTXDID( number modelId, number txdId )
bool engineResetModelTXDID( number modelId )
```

### CODE REVIEW

File: Client/game_sa/CModelInfoSA.cpp [hunk 1]
```diff
@@ -34,6 +34,7 @@ std::unordered_map<std::uint32_t, std::map<eVehicleDummies, CVector>> CModelInfo
 std::map<CTimeInfoSAInterface*, CTimeInfoSAInterface*>                CModelInfoSA::ms_ModelDefaultModelTimeInfo;
 std::unordered_map<DWORD, unsigned short>                             CModelInfoSA::ms_OriginalObjectPropertiesGroups;
 std::unordered_map<DWORD, std::pair<float, float>>                    CModelInfoSA::ms_VehicleModelDefaultWheelSizes;
+std::map<unsigned short, int>                                         CModelInfoSA::ms_DefaultTxdIDMap;
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.cpp:37]: You should consider using a fixed-size array, it's faster, and uses less memory.

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 2]
```diff
@@ -1256,7 +1258,34 @@ int CLuaEngineDefs::EngineRemoveShaderFromWorldTexture(lua_State* luaVM)
 
 uint CLuaEngineDefs::EngineGetModelTXDID(uint uiModelID)
 {
-    return g_pGame->GetRenderWare()->GetTXDIDForModelID(uiModelID);
+    CModelInfo* pModelInfo = g_pGame->GetModelInfo(uiModelID);
+
+    if (uiModelID >= 20000 || !pModelInfo)
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp]: Don't use magic numbers [Use an existing function or something of similar nature]

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 3]
```diff
@@ -1256,7 +1258,34 @@ int CLuaEngineDefs::EngineRemoveShaderFromWorldTexture(lua_State* luaVM)
...
+        throw std::invalid_argument("Expected a valid model ID in range [0-19999] at argument 1");
+
+    return pModelInfo->GetTextureDictionaryID();
+}
+
+bool CLuaEngineDefs::EngineSetModelTXDID(uint uiModelID, unsigned short usTxdId)
+{
+    CModelInfo* pModelInfo = g_pGame->GetModelInfo(uiModelID);
+
+    if (uiModelID >= 20000 || !pModelInfo)
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp]: Ditto

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 4]
```diff
@@ -1256,7 +1258,34 @@ int CLuaEngineDefs::EngineRemoveShaderFromWorldTexture(lua_State* luaVM)
...
+
+    pModelInfo->SetTextureDictionaryID(usTxdId);
+    return true;
+}
+
+bool CLuaEngineDefs::EngineResetModelTXDID(uint uiModelID)
+{
+    CModelInfo* pModelInfo = g_pGame->GetModelInfo(uiModelID);
+
+    if (uiModelID >= 20000 || !pModelInfo)
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp]: Ditto

**Review by Pirulax — CHANGES_REQUESTED** [review]

**Comment by CrosRoad95** [comment]
see how for example "EngineStreamingFreeUpMemory" lua function is implemented, move y our implementation above "Loadfunctions" so you can get rid of 2 lines of code in CLuaEngineDefs.h

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp (hunk 2 above)
Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp]: It is only draft =)
Will be changed

File: Client/game_sa/CModelInfoSA.cpp (hunk 1 above)
Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.cpp:37]: Fixed-size array will use at least 320 KB memory

Comment (Pirulax) [reply Client/game_sa/CModelInfoSA.cpp:37]: Memory fragmentation >>> usage

Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.cpp:37]: This struct will be empty on 95% servers. I see no reason to optimize this struct with memory cost for now.

**Comment by Pirulax** [comment]
I don't see you increasing the ref count of the TXDs anywhere, you should do that

**Comment by TheNormalnij** [comment]
> I don't see you increasing the ref count of the TXDs anywhere, you should do that

@Pirulax
Are your sure? This will prevent txd from unloading

**Comment by Pirulax** [comment]
A model [when loaded] will increase the refcount of a txd, and decrease it when it's unloaded.
So, when changing the txd's of a model you should manually unref the previous txd, and ref the new one.

**Comment by TheNormalnij** [comment]
> A model [when loaded] will increase the refcount of a txd, and decrease it when it's unloaded.
> So, when changing the txd's of a model you should manually unref the previous txd, and ref the new one.

Makes sense. Thanks you

**Review by Pirulax — APPROVED** [review]
lgtm

**Comment by Xenius97** [comment]
I've tested `engineSetModelTXDID` and `engineRequestTXD`, look good to me, but I noticed a possible memory leak. When you restart the resource, the requested ID won't be freed anymore (maybe if I manually call `engineFreeModel`). The same applies to `engineRequestTXD` and `engineFreeTXD`.

https://youtu.be/VoTWDzTKSjs
(Some default skins not working, See #3032)

**Comment by Pirulax** [comment]
Depends on #1802

**Comment by Pirulax** [comment]
Make sure to add the Wiki pages.

**Pull request merged**